src/hotspot/share/classfile/loaderConstraints.cpp

```diff
@@ -143,35 +215,41 @@ void LoaderConstraintTable::purge_loader_constraints() {
                             probe->loader_data(i)->loader_name_and_id());
             }
           }
-
-          continue;  // current element replaced, so restart without
-                     // incrementing n
-          }
-        n++;
+        }
       }
       // Check whether entry should be purged
       if (probe->num_loaders() < 2) {
-            if (lt.is_enabled()) {
-              ResourceMark rm;
-              lt.print("purging complete constraint for name %s",
-                         probe->name()->as_C_string());
-            }
+        if (lt.is_enabled()) {
+          ResourceMark rm;
+          lt.print("purging complete constraint for name %s",
+                   name->as_C_string());
+        }
 
-        // Purge entry
-        *p = probe->next();
-        FREE_C_HEAP_ARRAY(oop, probe->loaders());
-        free_entry(probe);
+        entry.remove_constraint(probe);
       } else {
 #ifdef ASSERT
         if (probe->klass() != NULL) {
           assert(probe->klass()->is_loader_alive(), "klass should be live");
         }
 #endif
-        // Go to next entry
-        p = probe->next_addr();
       }
     }
+    if (entry.num_constraints() == 0) {
+      // decrement name refcount before freeing
+      name->decrement_refcount();
+      entry.clean_up();
+      return true;
```

Comment:
This style of resource management in the hashtable entry is dangerous:
- On line 241 `entry._constraints` is a dangling pointer. You could explicitly set `entry._constraints` to NULL inside `clean_up()`, but that's extra code you need to write, and there's a chance you'd forget about doing it.
- If the table is deleted without you explicitly deleting the entries, you will have a memory leak.

I found the following pattern to be safer and require less work:

- Define a default constructor for LoaderConstraintEntry that zeros all fields
- Never instantiate a local LoaderConstraintEntry value and store that into the table. Instead use this:

```
LoaderConstraintEntry* p = put_if_absent(key, &created);
if (created) {
   p->initialize(....); // allocate  p->_constraints, etc.
}
```

- For the rule-of-three: delete the copy constructor and copy assignment operator. You don't need them. This way you can be sure you don't need to handle the copying of _constraints between two LoaderConstraintEntry objects.
- Free _constraints inside the LoaderConstraintEntry destructor.

This way, you know that _constraints is allocated in exactly one place, never copied, and deleted in exactly one place. And you will not have a memory leak.

Reply:
I was trying to avoid initialize functions, since that's what constructors do, but this usage would allow me to write a destructor, which is what I want.  This is actually quite nice.

Reply:
@iklam why is this pattern good? Why can't we create a properly initialized entry and add it to the map, deleting it if there was already one present? Doesn't this expose default initialized instances?

Reply:
There is a lock around all accesses to this table, so a partially constructed entry will never be exposed.